### PR TITLE
fix(rebuild): fixing child retirement during rebuild running

### PR DESCRIPTION
**Pull request by dsavitskiy**

### CODE REVIEW

**Comment by dsavitskiy** [comment]
bors try

File: io-engine/src/bdev/nexus/nexus_bdev_children.rs [hunk 1]
```diff
@@ -1068,7 +1069,7 @@ impl<'n> Nexus<'n> {
         // Cancel rebuild job for this child, if any.
         if let Some(job) = child.rebuild_job() {
             debug!("{self:?}: retire: stopping rebuild job...");
-            let terminated = job.terminate();
+            let terminated = job.terminate(RebuildOperation::Fail);
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nexus/nexus_bdev_children.rs]: would it be better to add a fail method to avoid having to pass operation and have assert inside?

**Review by tiagolobocastro — APPROVED** [review]

**Review by mtzaurus — APPROVED** [review]

File: io-engine/src/bdev/nexus/nexus_io_subsystem.rs [hunk 2]
```diff
@@ -174,7 +174,10 @@ impl<'n> NexusIoSubsystem<'n> {
 
                     let (s, r) = oneshot::channel::<i32>();
                     self.pause_waiters.push_back(s);
-                    r.await.unwrap();
+                    if r.await.is_err() {
+                        error!("NexusIoSubsystem is gone while waiting");
```
Comment (tiagolobocastro) [inline io-engine/src/bdev/nexus/nexus_io_subsystem.rs]: should we add self on these?

File: io-engine/src/core/device_monitor.rs [hunk 3]
```diff
@@ -65,12 +65,14 @@ pub async fn device_monitor_loop() {
...
                         Err(e) => {
-                            error!(
-                                "Failed to schedule removal request: {}",
-                                e.verbose()
-                            )
+                            error!("Failed to schedule removal request: {e}")
+                        }
+                        Ok(rx) => {
+                            if let Err(e) = rx.await {
+                                error!("Error processing removal request: {e}");
```
Comment (tiagolobocastro) [inline io-engine/src/core/device_monitor.rs]: isn't error just cancelled here? Might be more useful to add nexus name and child device?

File: io-engine/src/bdev/nexus/nexus_bdev_children.rs (hunk 1 above)
Comment (tiagolobocastro) [reply io-engine/src/bdev/nexus/nexus_bdev_children.rs]: @dsavitskiy wdyt?

**Review by tiagolobocastro — APPROVED** [review]

File: io-engine/src/bdev/nexus/nexus_io_subsystem.rs (hunk 2 above)
Comment (dsavitskiy) [reply io-engine/src/bdev/nexus/nexus_io_subsystem.rs]: self may be gone at this moment

**Comment by dsavitskiy** [comment]
bors try

File: io-engine/src/bdev/nexus/nexus_bdev_children.rs (hunk 1 above)
Comment (dsavitskiy) [reply io-engine/src/bdev/nexus/nexus_bdev_children.rs]: ok, will change

**Comment by dsavitskiy** [comment]
bors try

**Comment by dsavitskiy** [comment]
bors try

**Comment by dsavitskiy** [comment]
bors try

**Comment by dsavitskiy** [comment]
bors try

**Review by mtzaurus — APPROVED** [review]

**Comment by dsavitskiy** [comment]
bors merge

**Pull request merged**